authentication-scheme.mediawiki

```diff
@@ -0,0 +1,221 @@
+<pre>
+  NEP: <to be assigned>
+  Title: Authentication Scheme
+  Author: Erik Zhang <erik@neo.org>
+  Type: Standard
+  Status: Draft
+  Created: 2021-03-19
+</pre>
+
+==Abstract==
+
+This NEP describes a scheme for third-party applications to authenticate users based on NEO.
+The third-party applications mentioned here can be any type of application, including Dapps, websites, games, etc.
+
+==Motivation==
+
+Usually, an application saves user data on its own server. The authenticated user can modify the corresponding data through application-defined logic.
+Traditional Internet applications usually use an authentication scheme based on username and password, which is not suitable for blockchain.
+This authentication scheme allows third-party applications to authenticate users based on digital signatures.
+In this way, NEO users can log in to any third-party application that supports this scheme without registering, and allow applications to share user data on the blockchain.
+
+==Specification==
+
+===Payloads===
+
+The authentication process is divided into two steps.
+In the first step, the server sends a challenge payload to the client to request authentication.
+In the second step, the client sends the response payload to the server to complete the authentication.
+The payloads are transmitted in the form of JSON.
+
+====Challenge payload====
+
+<pre>
+{
+  "grant_type": "signature",
+  "allowed_algorithms": ["ECDSA-P256"],
+  "network": 5195086,
+  "nonce": "13458238842203010919",
+  "timestamp": 1616131368,
+  "callback": "https://someurl.com/callback"
+}
+</pre>
+
+=====grant_type=====
+
+This field should be fixed to <code>"signature"</code> and can be extended in the future to support more authentication schemes.
+
+=====allowed_algorithms=====
+
+An array containing the signature algorithms allowed by the server.
+
+The client should use one of the algorithms to sign the response data.
+If none of the algorithms are supported by the client, the authentication fails.
+
+=====network=====
+
+The magic number of the network.
+Indicates which network this authentication will be applied to.
+
+=====nonce=====
+
+A 64-bit unsigned integer used to identify this authentication.
+
+The server should record the expiration time of the nonce and check it when the response is received.
```

Comment:
This requires the authentication server to be stateful. That could be avoided by including a signature for the nonce and timestamp in the request (and requiring that the client provide the signature in the response).

Reply:
If the server is stateless, how to identify the clients?

Reply:
State will definitely be needed to provide any useful authorization (e.g. a table mapping from Neo address to roles and/or other application-specific identity data).  I'm just showing an opportunity to remove a database table and to avoid a database write on the initial authentication request (which could become a vector for unauthenticated denial-of-service).

Reply:
If the server does not record the nonce, it will become reusable. We need a mechanism to ensure that when a client is successfully authenticated, the nonce is immediately invalidated.

Reply:
It would be good to add mention here that the server is required to prevent nonce re-use:

> When the server receives the response, it should check whether the nonce has expired. If the nonce has expired or not exists, the authentication fails.

It's currently not clear that this is required and a server implementor might choose to delete a nonce after successful use rather than invalidate it.  I'm also curious as to why preventing nonce re-use is required (what security property breaks of the server doesn't do that?)

I'd still lean towards having the nonce invalidation happen post-authentication though.

Reply:
For example, I log in to a website by scanning a QR code. If the nonce can be reused, the second person who scans the code will kick my login.

Reply:
It sounds like the nonce is being used as a session ID? If that is the case there is a vulnerability:  Anyone who has observed the QR code can later hijack the session (it wouldn't be safe to use this flow in a room with security cameras, for example).  

To mitigate this, the nonce should be swapped for a session ID and this swap should only be allowed to happen once (note that there is still a race condition where an attacker can steal the session, though--to prevent that you need some sort of communication channel from the phone back to the browser).

Reply:
The nonce is a one-time random number. There should be a nonce to sessionId mapping table on the server side.

Reply:
> This requires the authentication server to be stateful. 

This is an interesting point @djnicholson, I need to study more on authentication. But perhaps, one way of not storing these nonces is to have some good pseudo-random algorithm, such as mersenne twister on server, starting with some secret seed, and then strongly hash mersenne twister outputs (via SHA-256, for example) to generate the nonce. The good thing is that, as long as some timestamp is involved, server can re-execute seeding process to find the nonce, instead of storing it. The bad thing is that, as soon as seed is leaked on server, nonces could be predicted.
Is it Neo servers or other servers that require being stateful?